include/osquery/http_client.h

```diff
@@ -0,0 +1,438 @@
+  boost::optional<std::string> remotePort() {
+    return uri_.port().size() ? uri_.port() : boost::optional<std::string>();
+  }
+
+  /// Returns the path part of a uri.
+  boost::optional<std::string> remotePath() {
+    return uri_.path().size() ? uri_.path() : boost::optional<std::string>();
+  }
+
+  /// Returns the protocol part of a uri. E.g. 'http' or 'https'
+  boost::optional<std::string> protocol() {
+    return uri_.scheme().size() ? uri_.scheme()
+                                : boost::optional<std::string>();
+  }
+
+  /// overloaded operator to set header of an http request
+  HTTP_Request& operator<<(const Header& h) {
+    this->T::set(h.name_, h.value_);
+    return *this;
+  }
+
+  /// uri can also be set via this method, useful for redirected request.
+  void uri(const std::string& url) {
+    uri_ = url;
+  }
+
+ private:
+  boost::network::uri::uri uri_;
+};
+
+/**
+ * @brief HTTP response class
+ *
+ * This class is inherited from implementation(boost.beast) http response class.
+ * This class gives convinent access to some functionality of implemention
+ * specific http response class.
+ *
+ */
+template <typename T>
+class HTTP_Response : public T {
+ public:
+  HTTP_Response() {}
+  HTTP_Response(const T& resp) : T(resp) {}
+
+  template <typename ITER>
+  class Iterator;
+  class Headers;
+
+  /// status of an http response
+  unsigned status() {
+    return this->T::result_int();
+  }
+
+  /// body of an http response
+  const std::string& body() {
+    return this->T::body;
+  }
+
+  /// All headers of an http response.
+  /// Headers can be accesed via HTTP_Response<T>::Iterator class
+  Headers headers() {
+    return Headers(this);
+  }
+};
+
+/**
+ * @brief HTTP response headers iterator class
+ *
+ */
+template <typename T>
+template <typename ITER>
+class HTTP_Response<T>::Iterator {
+ public:
+  Iterator(ITER iter) : iter_(iter) {}
+
+  Iterator operator++() {
+    ++iter_;
+    return *this;
+  }
+
+  Iterator operator++(int) {
+    Iterator tmp = *this;
+    ++*this;
+    return tmp;
+  }
+
+  bool operator!=(const Iterator& it) const {
+    return (iter_ != it.iter_);
+  }
+
+  auto operator-> () {
+    return std::make_shared<std::pair<std::string, std::string>>(
+        std::string(iter_->name_string()), std::string(iter_->value()));
+  }
+
+  auto operator*() {
+    return std::make_pair(std::string(iter_->name_string()),
+                          std::string(iter_->value()));
+  }
+
+ private:
+  ITER iter_;
+};
+
+/**
+ * @brief HTTP response headers helper class
+ *
+ * This class gives convenient access to all the
+ * headers of http response.
+ *
+ * e.g. -
+ * for (const auto& header : resp.headers()) {
+ *   header.first;
+ *   header.second;
+ * }
+ */
+template <typename T>
+class HTTP_Response<T>::Headers {
+ public:
+  Headers(T* resp) : resp_(resp) {}
+
+  std::string operator[](const std::string& name) {
+    return std::string(resp_->T::operator[](name));
+  }
```

Comment:
Why is this member function necesssary? It forces the caller to allocate a string for the `name` argument and it forces another allocation to construct `std::string` from `string_view`. This member function should be deleted.